api/v1alpha1/baseboardmanagement_types.go

```diff
@@ -17,29 +17,63 @@ limitations under the License.
 package v1alpha1
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// BaseboardManagementState represents the BaseboardManagement state.
+type BaseboardManagementState string
+
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
 // BaseboardManagementSpec defines the desired state of BaseboardManagement
 type BaseboardManagementSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+	// Host is the host IP address of the BaseboardManagement.
+	// +kubebuilder:validation:MinLength=1
+	Host string `json:"host"`
+
+	// AuthSecretRef is the SecretReference that contains authentication information of the BaseboardManagement.
+	// The Secret must contain username and password keys.
+	AuthSecretRef corev1.SecretReference `json:"authSecretRef"`
+
+	// Vendor is the vendor name of the BaseboardManagement.
+	// +kubebuilder:validation:MinLength=1
+	Vendor string `json:"vendor"`
+
+	// Power is the desired power state of the BaseboardManagement.
+	// +kubebuilder:validation:MinLength=1
+	Power string `json:"power"`
+
+	// Boot is the desired boot device for the BaseboardManagement.
+	Boot Boot `json:"boot"`
+}
+
+type Boot struct {
+	// BootDevice is the desired boot for the BaseboardManagement.
+	// +kubebuilder:validation:MinLength=1
+	BootDevice string `json:"bootDevice"`
+
+	// Persistent if True, boot device is permanently set for the BaseboardManagement.
+	// If False, boot device is set as one time boot.
+	// +kubebuilder:default=false
+	Persistent bool `json:"persistent,omitempty"`
 
-	// Foo is an example field of BaseboardManagement. Edit baseboardmanagement_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// EfiBoot specifies to EFI boot for the BaseboardManagement.
+	// +kubebuilder:default=false
+	EfiBoot bool `json:"efiBoot,omitempty"`
 }
 
 // BaseboardManagementStatus defines the observed state of BaseboardManagement
 type BaseboardManagementStatus struct {
```

Comment:
I'm not sure a status makes sense with BMC interactions. As we talked about, BMC interactions are more like K8s jobs. Their status is typically "completed" with "successful" or "failed". 

The fields you have in this status suggest a desired state model where the controller works to _continuously_ correct any state disparity. Combined with a `persistent` flag in the `Boot` type I'm not sure that really makes sense.

Desired state _may_ make sense if we want to describe a device that's permanently in a PXE boot state but that context is entirely about _persistent_ state not temporary states like on time PXE boot (note you could do temporary states with desired state modelling but I think it'll get more convoluted).

Finally, have we decided what will envelope the higher level of abstraction. Namely things like `MachineProvisioningState` and `MachineInUse` state. The `Tink` controller could envelope those states and we leave the BMC controller to just execute 1 off jobs.

Reply:
I think a power status displaying the present state is ok, it tells humans and controllers what the current real world state is.  

I think you are missing some fields you'll want here like
* Error message if controller can't connect 
* BMC info like
  * Reported BMC version
  * (future use) Redfish-provided hardware information (CPU, memory, disks, TPM, etc)

Reply:
> I'm not sure a status with these fields makes sense with BMC interactions. As we talked about, BMC interactions are more like K8s jobs. Their status is typically "completed" with "successful" or "failed".

So i believe the idea is `BaseboardManagement` CR is going to model the state of the BMC and we are going to have the controller reconcile and bring the BMC to the desired state.

We have another CRD, namely `PowerCycleJob` that would be similar to k8s jobs and would have the "completed", "success" status etc. My current idea is to have the Job CR have a list of actions but I am yet to figure out the details. Also, I think we have not discussed in detail about the high level abstractions? We can take that offline and discuss 👍

Reply:
> Edit, I just noticed your comment on `Boot`. A power only status makes sense, its the one-time boot aspect I'm not sure should be modeled with a status. If we want to track the persisted boot status that'd make sense also but we'd need to clearly define how we represent one-time boot settings.

My idea here was that the `Boot` field would indicate the desired first boot dev. 
In retrospect I realize that might be a bit confusing. I also noticed that bmc-lib does not have methods to check the boot dev it can only set them. cc @jacobweinstock

We can maybe have the `status` just track the persistent boot dev if set through a job?

Reply:
> Ideally you'd want to model these separately

Right. So I think the model we have decided to go with for now is we are not going to model Boot devices as a desired state. Only set boot device from "Actions" job. 
But we want to have the controller reconcile the desired power state.

> If you can create issues in bmclib for the methods you're looking for, I can help with those.

Thank you 😄  That would be super helpful. I will get started on that.